### PR TITLE
feat(devnet): Local Gossamer Devnet

### PR DESCRIPTION
**Pull request by timwu20**
## Changes

- docker-compose.yml  and associated docker images to setup a local devnet in a containerised environment.
- prometheus metrics are automatically piped to Datadog
- README.md for setup and running instructions

## Tests

```
go test devnet/
```

## Issues

- #1823 

## Primary Reviewer

- @noot

### CODE REVIEW

**Comment by noot** [comment]
@timwu20  looks epic, will test this out tomorrow!! do we want to add this to CI? (or wait until it's merged?)

File: devnet/README.md
Comment (kishansagathiya) [inline devnet/README.md:23]: This doesn't seem right (creating link at the same path). Also, root already has a `docker-compose.yml`. Is this a mistake or am I missing something?

Comment (timwu20) [reply devnet/README.md]: Good catch.  Updated the README.

**Comment by timwu20** [comment]
> @timwu20 looks epic, will test this out tomorrow!! do we want to add this to CI? (or wait until it's merged?)

I'm still working on the actual ECS integration using the docker-compose ECS plugin.  I'll integrate the CI then.  I thought there would be value for the team to be able to run it locally first.  The `docker-compose.yml` file with have more aws specific annotations and changes for it to deploy to ECS.

File: devnet/alice.Dockerfile
Comment (qdm12) [inline devnet/alice.Dockerfile:5]: I think this is enabled by default now

```suggestion
```

File: devnet/alice.Dockerfile
Comment (qdm12) [inline devnet/alice.Dockerfile:20]: You should cache Go dependencies with something similar to

https://github.com/ChainSafe/gossamer/blob/development/Dockerfile#L27-L35

File: devnet/alice.Dockerfile
Comment (qdm12) [inline devnet/alice.Dockerfile:33]: Since `EXPOSE` is for documentation only, let's precise the protocol as well :wink: 

```suggestion
EXPOSE 7001/tcp 8545/tcp 8546/tcp 8540/tcp 9876/tcp

```

**Comment by noot** [comment]
@timwu20 readme is a bit confusing, for this step:

> To symlink to the repository root
```
# cd into your Gossamer root
cd ~/dev/gossamer
```


do you mean the `--basepath`? so I need to copy the docker files to the basepath?

File: devnet/README.md
Comment (timwu20) [reply devnet/README.md:23]: right now alice.Dockerfile and bob.Dockerfile live in the `/devnet`, and I can't get the image to build without the codebase in the same directory.  Do you have any methods/suggestions around this?

Comment (timwu20) [reply devnet/README.md:23]: nevermind I got this working, and will remove this entire section.

File: devnet/alice.Dockerfile
Comment (timwu20) [reply devnet/alice.Dockerfile:32]: It is started already in the build, but I found it doesn't work without kicking it like this on container start.

File: cmd/gossamer/config.go
Comment (timwu20) [reply cmd/gossamer/config.go:683]: This gets parsed in `newHost`

File: devnet/README.md
Comment (qdm12) [reply devnet/README.md]: Nice :+1: Yes you can use the build context which can be different than where the Dockerfile is located.

File: devnet/alice.Dockerfile
Comment (qdm12) [reply devnet/alice.Dockerfile:32]: Maybe stop it in the build then, as soon as it gets auto-started? It feels kinda wrong to have something running in a 'frozen' image :sweat_smile:

**Comment by timwu20** [comment]
> @timwu20 readme is a bit confusing, for this step:
> 
> > To symlink to the repository root
> 
> ```
> # cd into your Gossamer root
> cd ~/dev/gossamer
> ```
> 
> do you mean the `--basepath`? so I need to copy the docker files to the basepath?

I'm getting rid of this whole step.  I got it working by just running `docker-compose up` in the devnet folder.

File: devnet/README.md
Comment (timwu20) [reply devnet/README.md:28]: In Datadog they call them tags, and that's how they are defined the datadag confd.yml.

File: devnet/alice.Dockerfile
Comment (timwu20) [reply devnet/alice.Dockerfile:29]: I'll try that.

File: devnet/docker-compose.yml
Comment (qdm12) [inline devnet/docker-compose.yml:8]: Can't the program auto-detect it's LAN IP address?

File: devnet/docker-compose.yml
Comment (qdm12) [inline devnet/docker-compose.yml:14]: nit no need for double quotes :wink: 

```suggestion
      - 7001
      - 8545
      - 8546
      - 8540
      - 9876
```

File: devnet/docker-compose.yml
Comment (qdm12) [inline devnet/docker-compose.yml:24]: nit maybe/ideally we should load that variable from a .gitignore'd `.env` file since exporting in your shell will leave a trace in your shell history and this is an API key. Or not if we don't care, that works too :smile_cat:

File: devnet/docker-compose.yml
Comment (qdm12) [inline devnet/docker-compose.yml:37]: When do we know when Alice is ready? Maybe should we have a shell/Go entrypoint to wait for Alice to be ready for bob and charlie? Or will e.g. `bob` fail if `alice` is not up yet?

File: dot/network/host.go
Comment (qdm12) [inline dot/network/host.go:69]: Define an error at global scope e.g. `var ErrPublicIPIsNotValid = errors.New("...")` and have this as

```suggestion
			return nil, fmt.Errorf("%w: %s", ErrPublicIPIsNotValid, cfg.PublicIP)
```

File: dot/network/host_test.go
Comment (qdm12) [inline dot/network/host_test.go:57]: Please precise the linter you want to silence with `//nolint:lintername`

File: dot/network/host_test.go
Comment (qdm12) [inline dot/network/host_test.go:70]: nit shorter, and clarified comment

```suggestion
		if i == len(addrInfo.Addrs) - 1 {
			// the last address is the address injected from Config.PublicIP so it's not blocked.
			require.True(t, privateIPs.AddrBlocked(addr))
			continue
		}
		require.False(t, privateIPs.AddrBlocked(addr))
```

File: go.mod
Comment (qdm12) [inline go.mod:50]: Can you merge this with the block above?

File: devnet/README.md
Comment (qdm12) [reply devnet/README.md]: Ah ok, Prometheus calls them labels that's why I was wondering. Maybe add `(Prometheus "labels")` after `additional tags`

**Comment by noot** [comment]
@timwu20 I'm having an issue running the commands:
```
r$ docker-compose up --abort-on-exit --build 
Builds, (re)creates, starts, and attaches to containers for a service.

Unless they are already running, this command also starts any linked services.

The `docker-compose up` command aggregates the output of each container. When
the command exits, all containers are stopped. Running `docker-compose up -d`
starts the containers in the background and leaves them running.

If there are existing containers for a service, and the service's configuration
or image was changed after the container's creation, `docker-compose up` picks
up the changes by stopping and recreating the containers (preserving mounted
volumes). To prevent Compose from picking up changes, use the `--no-recreate`
flag.

If you want to force Compose to stop and recreate all containers, use the
`--force-recreate` flag.
```

seems like the command is not working for me? :(

File: devnet/alice.Dockerfile
Comment (timwu20) [reply devnet/alice.Dockerfile:6]: this was essentially copy pasted from datadog instructions.  The only reason I made DD_API_KEY an ARG cause it's user dependant.

File: devnet/alice.Dockerfile
Comment (timwu20) [reply devnet/alice.Dockerfile:3]: just at build.

File: devnet/cmd/update-dd-agent-confd/main.go
Comment (timwu20) [reply devnet/cmd/update-dd-agent-confd/main.go:33]: I actually really like naked returns.  Check out the scale package 😛

File: devnet/cmd/update-dd-agent-confd/main.go
Comment (timwu20) [reply devnet/cmd/update-dd-agent-confd/main.go]: I've used the embed package before and found it a little cumbersome. I'll try it again.

File: devnet/cmd/update-dd-agent-confd/main_test.go
Comment (timwu20) [reply devnet/cmd/update-dd-agent-confd/main_test.go:53]: this was codegen'd from vscode.

File: devnet/docker-compose.yml
Comment (timwu20) [reply devnet/docker-compose.yml:8]: We use a library called `pubip` which requests from the internet from a bunch of those "what's my ip" services.  The `pubip` key I introduced forces the node to broadcast from the specified address.  I had some code that scanned all network interfaces, but if you have multiple addresses which one would you pick?

File: devnet/docker-compose.yml
Comment (timwu20) [reply devnet/docker-compose.yml:24]: I don't think it matters that much for people working locally on their computer.  I need to ensure that we don't broadcast it on the github actions logs when I deploy to ECS.

File: devnet/docker-compose.yml
Comment (timwu20) [reply devnet/docker-compose.yml:37]: In @noot PR for the `--babe-lead` flag, she changed the behaviour for all other non-lead auth nodes to wait for the first block.  So we still wait for alice to start up first, but it's not completely necessary.

File: dot/network/host.go
Comment (timwu20) [reply dot/network/host.go:69]: The other errors for the host are produced with `fmt.Errorf`.  I don't mind doing this, but we should be changing all the errors for the node.

**Comment by timwu20** [comment]
> @timwu20 I'm having an issue running the commands:
> 
> ```
> r$ docker-compose up --abort-on-exit --build 
> Builds, (re)creates, starts, and attaches to containers for a service.
> 
> Unless they are already running, this command also starts any linked services.
> 
> The `docker-compose up` command aggregates the output of each container. When
> the command exits, all containers are stopped. Running `docker-compose up -d`
> starts the containers in the background and leaves them running.
> 
> If there are existing containers for a service, and the service's configuration
> or image was changed after the container's creation, `docker-compose up` picks
> up the changes by stopping and recreating the containers (preserving mounted
> volumes). To prevent Compose from picking up changes, use the `--no-recreate`
> flag.
> 
> If you want to force Compose to stop and recreate all containers, use the
> `--force-recreate` flag.
> ```
> 
> seems like the command is not working for me? :(

Seems odd.  What version of Docker are you running? What happens if you omit the `--build` flag?

File: devnet/alice.Dockerfile
Comment (qdm12) [reply devnet/alice.Dockerfile:3]: Ok let's not have `ENV DD_API_KEY=${DD_API_KEY}` then. `ENV` is mostly to persist it in the final image for runtime usage.

File: devnet/docker-compose.yml
Comment (qdm12) [reply devnet/docker-compose.yml:37]: Awesome, let's leave those `depends_on` for documentation really then :wink:

File: dot/network/host.go
Comment (qdm12) [reply dot/network/host.go:69]: Not really, just do it bit by bit I think, especially since it's new code. I don't believe we should go all in and refactor all errors returned :sweat_smile:

File: go.mod
Comment (timwu20) [reply go.mod:50]: this is what `go mod tidy` does automatically and I'm not sure why.  I'll try it and see if it complains.

File: devnet/README.md
Comment (noot) [inline devnet/README.md:40]:
```suggestion
### Dockerfiles
```

**Review by noot — APPROVED** [review]
worked for me - great work!

File: devnet/cmd/update-dd-agent-confd/main_test.go
Comment (timwu20) [reply devnet/cmd/update-dd-agent-confd/main_test.go:45]: When I add a test case, that fails, both tests show failed for some reason.  I'm going to leave out the `t.Parallel()`.

```
=== RUN   Test_marshalYAML
=== RUN   Test_marshalYAML/zero_case_options
=== PAUSE Test_marshalYAML/zero_case_options
=== RUN   Test_marshalYAML/options_with_ns_and_tags
=== PAUSE Test_marshalYAML/options_with_ns_and_tags
=== CONT  Test_marshalYAML/zero_case_options
=== CONT  Test_marshalYAML/options_with_ns_and_tags
=== CONT  Test_marshalYAML/zero_case_options
    /Users/timothywu/dev/gossamer/devnet/cmd/update-dd-agent-confd/main_test.go:82: marshalYAML() diff =   bytes.Join({
          	"init_config: {}\ninstances:\n- prometheus_url: http://127.0.0.1:98",
          	"76/metrics\n  namespace: SomeNamespace\n  metrics:\n  - gossamer",
        + 	"1",
          	"_*\n  - network_*\n  - service_*\n  - system_*\n  health_service_che",
          	"ck: true\n  tags:\n  - some\n  - tags\n",
          }, "")
=== CONT  Test_marshalYAML/options_with_ns_and_tags
    /Users/timothywu/dev/gossamer/devnet/cmd/update-dd-agent-confd/main_test.go:82: marshalYAML() diff =   bytes.Join({
          	"init_config: {}\ninstances:\n- prometheus_url: http://127.0.0.1:98",
          	"76/metrics\n  namespace: SomeNamespace\n  metrics:\n  - gossamer",
        + 	"1",
          	"_*\n  - network_*\n  - service_*\n  - system_*\n  health_service_che",
          	"ck: true\n  tags:\n  - some\n  - tags\n",
          }, "")
--- FAIL: Test_marshalYAML (0.00s)
    --- FAIL: Test_marshalYAML/zero_case_options (0.00s)
    --- FAIL: Test_marshalYAML/options_with_ns_and_tags (0.00s)
```

File: dot/network/host.go
Comment (timwu20) [reply dot/network/host.go:69]: `host` is a private struct.  Not sure this is really required, since the only thing calling `newHost` is `NewService` and even that doesn't have package errors at this point.

File: devnet/docker-compose.yml
Comment (qdm12) [reply devnet/docker-compose.yml:8]: For LAN ip addresses, you can `go get github.com/vishvananda/netlink` and use

```go
func Test_findLANIP(t *testing.T) {
	ip, err := findLANIP()
	require.NoError(t, err)
	t.Log(ip)
}

var (
	ErrNoLinkFound            = errors.New("no link was found")
	ErrDefaultGatewayNotFound = errors.New("default gateway not found")
)

func findLANIP() (ip net.IP, err error) {
	links, err := netlink.LinkList()
	if err != nil {
		return nil, fmt.Errorf("cannot list links: %w", err)
	} else if len(links) == 0 {
		return nil, ErrNoLinkFound
	}

	for _, link := range links {
		routes, err := netlink.RouteList(link, netlink.FAMILY_ALL)
		if err != nil {
			return nil, fmt.Errorf(
				"cannot list routes for link %d: %w",
				link.Attrs().Index, err)
		}

		for _, route := range routes {
			if route.Dst == nil {
				return route.Gw, nil
			}
		}
	}

	return nil, ErrDefaultGatewayNotFound
}
```

Although that's limited to unix systems and introduces an indirect dependency on syscall (so for windows we would need different build tags etc.)

File: devnet/docker-compose.yml
Comment (qdm12) [reply devnet/docker-compose.yml:24]: I think you can also use `secrets:` in docker-compose.yml, if that may be of use. [LINK](https://docs.docker.com/compose/compose-file/compose-file-v3/#secrets)

File: devnet/cmd/update-dd-agent-confd/main.go
Comment (qdm12) [inline devnet/cmd/update-dd-agent-confd/main.go:12]: Can't you use the stdlib's `flag` package for this?

File: devnet/bob.Dockerfile
Comment (timwu20) [reply devnet/bob.Dockerfile:33]: I don't see a case where we do want to adjust the bootnodes for the `gssmr` devnet.

**Comment by timwu20** [comment]
I don't know why I can't reply to your feedback about the lan IP stuff.  Since this will only ever be used by the devnet, I'd rather not bring in extra code when the user can just force the node to broadcast the intended address.

File: devnet/cmd/update-dd-agent-confd/main.go
Comment (timwu20) [reply devnet/cmd/update-dd-agent-confd/main.go:12]: I prefer `go-flags` and just using a struct.  It has served me well for previous projects.

File: dot/network/host.go
Comment (kishansagathiya) [inline dot/network/host.go:79]: should we return error here?

If we are here, that means we are failing to set `externalAddr`. `externalAddr` is getting used in the code after this, but not getting set. Looks like we should return the error here.

File: dot/network/host.go
Comment (kishansagathiya) [inline dot/network/host.go:87]: We could use if, else to just set ip and shift some common code out side those blocks

```go
	var ip net.IP
	if cfg.PublicIP != "" {
		ip = net.ParseIP(cfg.PublicIP)
		if ip == nil {
			return nil, fmt.Errorf("invalid public ip: %s", cfg.PublicIP)
		}
		logger.Debugf("using config PublicIP: %s", ip)

	} else {
		ip, err = pubip.Get()
		if err != nil {
			return nil, fmt.Errorf("failed to get public IP error: %v", err)
		}
		logger.Debugf("got public IP", "IP", ip)
	}

	externalAddr, err := ma.NewMultiaddr(fmt.Sprintf("/ip4/%s/tcp/%d", ip, cfg.Port))
	if err != nil {
		return nil, err
	}
```
(Looks like, code suggestions are not supported for more that a few lines)

File: devnet/bob.Dockerfile
Comment (qdm12) [reply devnet/bob.Dockerfile:19]: Yes and also we should start and stop the service here, no point letting it run I think

File: dot/network/host.go
Comment (timwu20) [reply dot/network/host.go:79]: This is how the code was.  In the event we can't get a public IP via `pubip` I would assume nothing gets broadcasted.  What do you think @noot?

File: devnet/bob.Dockerfile
Comment (timwu20) [reply devnet/bob.Dockerfile:19]: I modified this in my other branch to not start it here.  I'll add it to this branch.

File: dot/network/host.go
Comment (timwu20) [reply dot/network/host.go:87]: I'm adding another flag for `cfg.PublicDNS` so this is changing in the next PR, to have an `else if` case where the multiaddr is no longer `"/ip4/%s.."`, but rather `"/dns/%s..."`.

File: devnet/bob.Dockerfile
Comment (timwu20) [reply devnet/bob.Dockerfile:19]: in terms of readability, does it make sense to append multiple RUN statements together?

File: devnet/bob.Dockerfile
Comment (timwu20) [reply devnet/bob.Dockerfile:31]: Move them up where? If I move the `key` and `pubip` args/env above the `RUN go install` doesn't that break caching for subsequent builds?

File: devnet/bob.Dockerfile
Comment (kishansagathiya) [reply devnet/bob.Dockerfile:19]: I don't think readability would worsen if we append statements. You could also indent it to make to more readable by writing each command on different lines using `\\`

File: devnet/bob.Dockerfile
Comment (kishansagathiya) [reply devnet/bob.Dockerfile:31]: I don't have a good understanding of how caching works in docker, will read up and comment again! I will resolve this for now and un-resolve if I find any issues.

File: devnet/bob.Dockerfile
Comment (timwu20) [reply devnet/bob.Dockerfile:19]: I concatenated the RUN lines together and it made no significant difference in build time.

**Review by kishansagathiya — APPROVED** [review]

File: devnet/bob.Dockerfile
Comment (qdm12) [reply devnet/bob.Dockerfile:33]: At least the IP address `10.5.0.2` could be changed if we change the docker network right?

**Comment by qdm12** [comment]
> I don't know why I can't reply to your feedback about the lan IP stuff.

It's because it's replies to conversations above (probably hidden since they are many conversations) :wink:

File: dot/network/host.go
Comment (noot) [reply dot/network/host.go:79]:
```
		libp2p.AddrsFactory(func(as []ma.Multiaddr) []ma.Multiaddr {
			addrs := []ma.Multiaddr{}
			for _, addr := range as {
				if !privateIPs.AddrBlocked(addr) {
					addrs = append(addrs, addr)
				}
			}
			if externalAddr == nil {
				return addrs
			}

			return append(addrs, externalAddr)
		}),
```

other addresses can still be advertised even if `pubip.Get` fails, so I think that's why I didn't make it return an error

File: devnet/bob.Dockerfile
Comment (timwu20) [reply devnet/bob.Dockerfile:33]: In the next PR I do actually have to change the bootnode ports, so port will be configurable on `bob.Dockerfile`.  I'm trying to hardcode the configuration for `alice.Dockerfile` since that is the lead node.  I think being explicit in this case  shows that everything is dependant on alice.

**Pull request merged**
